src/dd/StateGeneration.cpp

```diff
@@ -0,0 +1,145 @@
+
+#include <algorithm>
+#include <array>
+#include <cassert>
+#include <cmath>
+#include <complex>
+#include <cstddef>
+#include <limits>
+#include <random>
+#include <utility>
+#include <vector>
+
+namespace dd {
+namespace {
+using Generator = std::mt19937_64;
+using AngleDistribution = std::uniform_real_distribution<double>;
+using IndexDistribution = std::uniform_int_distribution<std::size_t>;
+
+/**
+ * @brief Return random complex number on the unit circle, such that,
+ *        its squared magnitude is one.
+ */
+std::complex<double> randomComplexOnUnitCircle(Generator& gen,
+                                               AngleDistribution& dist) {
+  const double angle = dist(gen);
+  return {std::cos(angle), std::sin(angle)};
+}
+
+/**
+ * @brief Generate random node with terminal edges initialized with random
+ * weights. The function ensures that the norm of the edge weights is 1.
+ */
+vEdge randomNode(Qubit v, Generator& gen, AngleDistribution& dist,
+                 Package& dd) {
+  constexpr double eps = std::numeric_limits<double>::epsilon();
+
+  const auto alpha = randomComplexOnUnitCircle(gen, dist);
+  const auto beta = randomComplexOnUnitCircle(gen, dist);
+  const auto norm = std::sqrt(2);
+
+  const std::array<vEdge, RADIX> edges{
+      vEdge::terminal(dd.cn.lookup(alpha / norm)),
+      vEdge::terminal(dd.cn.lookup(beta / norm)),
+  };
+
+  // Check: Properly normalized.
+  assert(std::sqrt(ComplexNumbers::mag2(edges[0].w) +
+                   ComplexNumbers::mag2(edges[1].w)) < 1 + eps);
+
+  return dd.makeDDNode(v, edges);
+}
+} // namespace
+
+VectorDD generateExponentialState(const std::size_t levels, Package& dd) {
+  std::random_device rd;
+  return generateExponentialState(levels, rd(), dd);
+}
+
+VectorDD generateExponentialState(const std::size_t levels,
+                                  const std::size_t seed, Package& dd) {
+  std::vector<std::size_t> nodesPerLevel(levels - 1); // [2, 4, 8, ...]
+  std::generate(nodesPerLevel.begin(), nodesPerLevel.end(),
+                [exp = 1]() mutable { return std::pow(2, exp++); });
+  return generateRandomState(levels, nodesPerLevel, ROUNDROBIN, seed, dd);
+}
+
+VectorDD generateRandomState(const std::size_t levels,
+                             const std::vector<std::size_t>& nodesPerLevel,
+                             const GenerationWireStrategy strategy,
+                             Package& dd) {
+  std::random_device rd;
+  return generateRandomState(levels, nodesPerLevel, strategy, rd(), dd);
+}
+
+VectorDD generateRandomState(const std::size_t levels,
+                             const std::vector<std::size_t>& nodesPerLevel,
+                             const GenerationWireStrategy strategy,
+                             const std::size_t seed, Package& dd) {
+  assert(levels > 0U && "Number of levels must be greater than zero");
+  assert(nodesPerLevel.size() == levels - 1 &&
+         "Number of levels - 1 must match nodesPerLevel size");
```

Comment:
I believe this method is also missing the safety checks for the node configuration from the original branch. Specifically that there can never be more than twice as many nodes as in the previous level. The condition each level can have at most a power of two in nodes is implied by that given how there is only one root node.